### PR TITLE
added support for custom lifecycles

### PR DESCRIPTION
**Pull request by musselwhizzle**

### CODE REVIEW

File: library/src/main/java/com/climate/mirage/Mirage.java [hunk 1]
```diff
@@ -272,6 +273,12 @@ public MirageTask go(MirageRequest request) {
             return null;
         }
 
+        // exit early if the lifecycle is not active.
+        Lifecycle lifecycle = request.lifecycle();
+        if (lifecycle != null && lifecycle.getCurrentState() == Lifecycle.State.DESTROYED) {
+        	return null;
+		}
```
Comment (AnkitPatanaik) [inline library/src/main/java/com/climate/mirage/Mirage.java]: misaligned brace

Comment (AnkitPatanaik) [inline library/src/main/java/com/climate/mirage/Mirage.java:280]: misaligned brace

**Review by vijaygandhi — APPROVED** [review]

File: library/src/main/java/com/climate/mirage/requests/MirageRequest.java [hunk 2]
```diff
@@ -107,6 +110,7 @@ public void recycle() {
 			resizeProcessor.setDimensions(0, 0, ResizeProcessor.STRATEGY_SCALE_FREE);
 		}
         bitmapProvider = null;
+		lifecycle = null;
```
Comment (AnkitPatanaik) [inline library/src/main/java/com/climate/mirage/requests/MirageRequest.java:113]: indentation is off here

File: library/src/main/java/com/climate/mirage/tasks/MirageTask.java [hunk 3]
```diff
@@ -86,5 +101,14 @@ private void internalCancel() {
...
         if (request.target() != null) request.target().onCancel();
         callback.onCancel(this, request);
+		removeLifecycleObserver();
     }
+
+    private void removeLifecycleObserver() {
+    	if (lifecycle != null) {
+    		lifecycle.removeObserver(lifecycleObserver);
+		}
+    	lifecycle = null;
```
Comment (AnkitPatanaik) [inline library/src/main/java/com/climate/mirage/tasks/MirageTask.java:111]: why does the lifecycle have to become null here?

File: library/src/test/java/com/climate/mirage/tasks/MirageTaskTest.java [hunk 4]
```diff
@@ -29,7 +34,11 @@
 
 import java.io.ByteArrayOutputStream;
 import java.io.IOException;
+import java.io.InterruptedIOException;
 import java.net.URL;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ThreadPoolExecutor;
```
Comment (AnkitPatanaik) [inline library/src/test/java/com/climate/mirage/tasks/MirageTaskTest.java]: it seems like this file only had a bunch of imports added?

**Review by AnkitPatanaik — APPROVED** [review]

**Pull request merged**